### PR TITLE
Amazon GuardDuty supports exporting findings to an Amazon S3 bucket #10920

### LINKED ISSUE #10920: Amazon GuardDuty supports exporting findings to an Amazon S3 bucket
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Amazon GuardDuty customers can now export findings to Amazon S3.

### New or Affected Resource(s)

* `aws_guardduty_publishing_destination`

### Potential Terraform Configuration

```hcl
resource "aws_guardduty_publishing_destination" "example" {}
```

### References

[Announcement](https://aws.amazon.com/about-aws/whats-new/2019/11/amazon-guardduty-supports-exporting-findings-to-an-amazon-s3-bucket/).
[User guide](https://docs.aws.amazon.com/guardduty/latest/ug/guardduty_exportfindings.html).
[API reference](https://docs.aws.amazon.com/guardduty/latest/APIReference/API_CreatePublishingDestination.html)

### PR DESCRIPTION
**Pull request by shaepe**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #10920

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
**New Resource**: `aws_guardduty_publishing_destination` (#10920)
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAwsGuardDutyPublishingDestination'

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -count 1 -parallel 20 -run=TestAccAwsGuardDutyPublishingDestination -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAwsGuardDutyPublishingDestination_basic
--- PASS: TestAccAwsGuardDutyPublishingDestination_basic (105.17s)
=== RUN   TestAccAwsGuardDutyPublishingDestination_import
--- PASS: TestAccAwsGuardDutyPublishingDestination_import (106.17s)
PASS
...
```

### CODE REVIEW

File: aws/resource_aws_guardduty_publishing_destination.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+				ForceNew: true,
+			},
+			"destination_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "S3",
+			},
+			"destination_arn": {
+				Type:     schema.TypeString,
+				Required: true,
```
Comment (ewbankkit) [inline aws/resource_aws_guardduty_publishing_destination.go]: Add `ValidateFunc: validateArn,`?

File: aws/resource_aws_guardduty_publishing_destination.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+			},
+			"kms_key_arn": {
+				Type:     schema.TypeString,
+				Required: true,
```
Comment (ewbankkit) [inline aws/resource_aws_guardduty_publishing_destination.go]: Add `ValidateFunc: validateArn,`?

File: aws/resource_aws_guardduty_publishing_destination.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+		Schema: map[string]*schema.Schema{
+			"detector_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"destination_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "S3",
```
Comment (ewbankkit) [inline aws/resource_aws_guardduty_publishing_destination.go]: Can use AWS SDK-provided constant `guardduty.DestinationTypeS3`.

File: aws/resource_aws_guardduty_publishing_destination.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+func guardDutyPublishingDestinationRefreshStatusFunc(conn *guardduty.GuardDuty, destinationID, detectorID string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		input := &guardduty.DescribePublishingDestinationInput{
+			DetectorId:    aws.String(detectorID),
+			DestinationId: aws.String(destinationID),
+		}
+		resp, err := conn.DescribePublishingDestination(input)
+		if err != nil {
+			return nil, "failed", err
```
Comment (ewbankkit) [inline aws/resource_aws_guardduty_publishing_destination.go]: Maybe add as

```go
// Constants not currently provided by the AWS Go SDK
const (
	guardDutyPublishingStatusFailed  = "FAILED"
)
```

at top of file?

File: aws/resource_aws_guardduty_publishing_destination.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+		}
+		return resp, *resp.Status, nil
+	}
+}
+
+func resourceAwsGuardDutyPublishingDestinationRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).guarddutyconn
+
+	destination_id, detector_id, err_state_read := decodeGuardDutyPublishDestinationID(d.Id())
```
Comment (ewbankkit) [inline aws/resource_aws_guardduty_publishing_destination.go]: Prefer camelCase over snake_case here (and others).

File: aws/resource_aws_guardduty_publishing_destination.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+		return err_state_read
+	}
+
+	input := guardduty.DeletePublishingDestinationInput{
+		DestinationId: aws.String(destination_id),
+		DetectorId:    aws.String(detector_id),
+	}
+
+	log.Printf("[DEBUG] Delete GuardDuty publishing destination: %s", input)
+	_, err := conn.DeletePublishingDestination(&input)
```
Comment (ewbankkit) [inline aws/resource_aws_guardduty_publishing_destination.go]: What happens if the PublishingDestination has been deleted outside Terraform?
Do we need the `if isAWSErr(err, guardduty.ErrCodeBadRequestException,...` check here?

**Comment by miguelaferreira** [comment]
Hi @shaepe.  Thanks for this PR, my team is looking to use this resource and I'm wondering if you will have the time to address the review from @ewbankkit, otherwise we could do that to make progress on your PR.

File: aws/resource_aws_guardduty_publishing_destination.go [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+			"detector_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"destination_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  guardduty.DestinationTypeS3,
+			},
```
Comment (ewbankkit) [inline aws/resource_aws_guardduty_publishing_destination.go:41]: Add
```go
ValidateFunc: validation.StringInSlice([]string{
  guardduty.DestinationTypeS3,
}, false),
```

File: aws/resource_aws_guardduty_publishing_destination.go [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Error waiting for GuardDuty PublishingDestination status to be \"%s\": %s",
+			guardduty.PublishingStatusPublishing, err)
+	}
+
+	d.SetId(fmt.Sprintf("%s:%s", d.Get("detector_id"), *output.DestinationId))
```
Comment (ewbankkit) [inline aws/resource_aws_guardduty_publishing_destination.go:89]:
```go
d.SetId(fmt.Sprintf("%s:%s", d.Get("detector_id"), aws.StringValue(output.DestinationId)))
```

File: aws/resource_aws_guardduty_publishing_destination.go [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+		DestinationType: aws.String(d.Get("destination_type").(string)),
+	}
+
+	log.Printf("[DEBUG] Creating GuardDuty publishing destination: %s", input)
+	output, err := conn.CreatePublishingDestination(&input)
+	if err != nil {
+		return fmt.Errorf("Creating GuardDuty publishing destination failed: %s", err.Error())
+	}
+
+	stateConf := &resource.StateChangeConf{
```
Comment (ewbankkit) [inline aws/resource_aws_guardduty_publishing_destination.go:75]: Could you please move this code to `aws/internal/service/guardduty/waiter` and follow the pattern there?
See https://github.com/terraform-providers/terraform-provider-aws/issues/12840.

**Comment by bendehaan** [comment]
@ewbankkit moved to new PR since original author is unresponsive: https://github.com/terraform-providers/terraform-provider-aws/pull/13894/

**Comment by ewbankkit** [comment]
Closing in favour of #13894.
@shaepe Thanks for your work on this.

**Pull request closed without merging**